### PR TITLE
Knack Banner integration

### PR DESCRIPTION
**Pull request by chiaberry**
Long awaited knack banner integration. Paired with Sergio on connecting to the shared drive, he led me to use https://pypi.org/project/smbprotocol/ and I updated the dockerfile with the extra requirements. 

The employee email file is named based on the date, for example `2021-07-12-ATDUsers.csv`. I asked Stephanie if we can get ctm to also include a consistently named filename. She said we could, I will follow up after this PR is approved about what the filename will be. (It's passed in as an environment variable)

A few other notes: 
- While testing, I found a number of records in knack missing employee ids. I checked with Diana and she said that should not happen moving forward, so I went into the the prod and test HR apps and manually updated the employee ids. Otherwise, records missing ids are marked as inactive. The records for those employees in banner won't match up correctly (since there is no pk) and we get a "unique email" error when they are trying to be added instead of updated. This script doesn't have a way to handle this scenario if it continues, beyond a user manually updating records based on the error_report.
- The first few times I ran this script knack timed out, because there were over 80 records to update. (The titles in banner did not match the titles in knack) I saw [a script](https://github.com/cityofaustin/atd-airflow/blob/master/dags/python_scripts/atd_mds_monthly_report.py) where Sergio reattempts every failed run until we get a 200. However, subsequent runs of the script only have had to update one or two accounts and have not timed out. 
- Knack doesn't like improperly formed emails, so I added a function to remove records that do not have emails. 

edited to add -- added `docker-image-build-push.yml` and secrets to repo

### CODE REVIEW

**Review by dianamartin — APPROVED** [review]

**Review by dianamartin — APPROVED** [review]

File: update_employees.py [hunk 1]
```diff
@@ -70,7 +80,61 @@ def get_employee_data():
...
+    """
+    Read csv from shared drive with employee emails
+    :return: dictionary of emails
+    """
+    employee_emails = {}
+
+    smbclient.ClientConfig(username=os.getenv("SHAREDDRIVE_USERNAME"), password=os.getenv("SHAREDDRIVE_PASSWORD"))
+
+    emails_csv = os.getenv("SHAREDDRIVE_FILEPATH")
+    with smbclient.open_file(emails_csv, mode='r') as emails:
```
Comment (johnclary) [inline update_employees.py:93]: wow—cool!

Comment (chiaberry) [reply update_employees.py:93]: my same reaction when Sergio showed me the python package

File: update_employees.py [hunk 2]
```diff
@@ -184,24 +300,27 @@ def main():
...
-    payload = build_payload(records_knack, records_mapped, pk_field, status_field)
-
     password_field = PASSWORD_FIELD[KNACK_APP_NAME]
-    set_passwords(payload, password_field)
+    email_field = EMAIL_FIELD[KNACK_APP_NAME]
+    payload = build_payload(records_knack, records_mapped, pk_field, status_field, password_field)
+    cleaned_payload = remove_empty_emails(payload, email_field)
 
-    print(f"{len(payload)} records to process.")
+    print(f"{len(cleaned_payload)} total records to process in Knack.")
```
Comment (johnclary) [inline update_employees.py:320]: since you have a logger configured, change all the `print()`s to logs?

**Review by johnclary — APPROVED** [review]
@chiaberry amazing work!! there are a lot of annoying constraints with these data sources and you've handled them really well. 

r.e. the question of re-trying on failed requests, since [fairly recently](https://github.com/cityofaustin/knackpy/releases/tag/1.0.17) knackpy will actually re-try automatically when you're making a request. it's sad that even 80 records can't be reliably updated, but i think the approach you've taken is fine.

🌬️ ⚙️ ⛵

**Pull request merged**